duckduckgo/BrowserServicesKit
# Improve handling lists in Sync

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1205884549503153/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2192
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1886
What kind of version bump will this require?: Major

**Optional**:

Tech Design URL:
CC:

**Description**:
Improve handling of lists in syncable models' payloads (which currently is limited to bookmarks
folders and favorites folders children). This change updates how lists are sent to Sync in PATCH
payloads by including current state as well as the set of removed and inserted items.
To allow for this, last sync payload is kept for all bookmarks folders in
`lastChildrenPayloadReceivedFromSync` property added on BookmarkEntity entity.
If the payload is not available (e.g. first sync), the legacy mechanism is used (i.e. sending only
the current contents of the list). The schema of a Sync response payload does not change so
the implementation handles it too.



**Steps to test this PR**:
1. Run the client app (iOS and/or macOS)
2. Import/populate bookmarks and enable Sync.
3. Move bookmarks across folders and verify that changes are properly synced between devices.



**OS Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
